applications/StructuralMechanicsApplication/custom_constraints/link_constraint.cpp

```diff
@@ -0,0 +1,335 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Máté Kelemen
+//
+
+// Project includes
+#include "custom_constraints/link_constraint.hpp" // LinkConstraint
+#include "includes/define.h"
+#include "includes/process_info.h"
+#include "includes/variables.h" // DISPLACEMENT_X, DISPLACEMENT_Y, DISPLACEMENT_Z
+#include "includes/serializer.h" // Serializer
+#include "includes/variables.h" // CONSTITUTIVE_MATRIX, INTERNAL_FORCES_VECTOR
+
+// STL includes
+#include <algorithm> // std::max_element, std::equal, std::transform
+
+
+namespace Kratos {
+
+
+struct LinkConstraint::Impl
+{
+    using ValueType = double;
+
+    static void ComputeRelationMatrix(LinkConstraint::MatrixType& rRelationMatrix,
+                                      LinkConstraint::MatrixType& rHessian,
+                                      LinkConstraint::VectorType& rConstraintGaps,
+                                      const std::array<ValueType,6>& rLastPositions,
+                                      const std::array<ValueType,6>& rLastDisplacements,
+                                      const unsigned Dimensions)
+    {
+        // Make sure we don't try to allocate an underflown unsigned integer.
+        KRATOS_ERROR_IF_NOT(Dimensions);
+
+        // Set output sizes.
+        rRelationMatrix.resize(1, 2 * Dimensions);
+        rHessian.resize(2 * Dimensions, 2 * Dimensions);
```

Comment:
Why woudn't you do the following
```suggestion
        noalias(rRelationMatrix) = ZeroMatrix(1, 2 * Dimensions);
        noalias(rHessian) = ZeroMatrix(2 * Dimensions, 2 * Dimensions);
```

This is more readable as well than using the `std::fill` in my opinion.

Reply:
Maybe, but that's **always** allocating new memory, while resizing does not if the size is already correct. I should have set `preserve = false` though.

Reply:
You're right, `ZeroMatrix` does not allocate.